### PR TITLE
Standardize inbound lw

### PR DESCRIPTION
**Pull request by ros3bud**
Theses changes reflect the process of reverse engineering the lendio ids to get their represented values. They include updating the tests to reflect the way the data is now being handled and the removal of the extra lendio fields which previously held ids. 

They also include the changes that reflect using the handler for typing.

### CODE REVIEW

**Comment by cgrayson** [comment]
I don't see an issue that details the changes needed here. You said they include "the process of reverse engineering the lendio ids to get their represented values", but I don't understand what that means. Can you provide more info on that?

**Comment by ros3bud** [comment]
Please see issue #54 for further details.

File: src/outbound.coffee [hunk 1]
```diff
@@ -50,7 +51,7 @@ request = (vars) ->
     if !name? or name == ''
       delete content[name]
     else
-      content[name] = value?.valueOf()
+      content[name] = value
```
Comment (cgrayson) [inline src/outbound.coffee]: I don't think this is necessary anymore. :smiley:

File: src/outbound.coffee [hunk 2]
```diff
@@ -202,15 +191,17 @@ fullName = (first, last) ->
   return name
 
 parseBool = (value) ->
-  return null unless value?
-  if value.valueOf() then 1 else 0
+  return null unless value
+  value = types.boolean.parse(value).valueOf()
+  return 1 if value
+  return 0
```
Comment (cgrayson) [inline src/outbound.coffee]: I don't see why the changes on lines 195-197 are needed. This function is used twice, both times with variables that are already `boolean`.

File: src/outbound.coffee [hunk 3]
```diff
@@ -202,15 +191,17 @@ fullName = (first, last) ->
...
 
 inRange = (value, min, max) ->
-  return false unless value?.valid
+  return false unless value.min or value.max or value.avg 
   (value.min >= min and value.max <= max) or (value.avg >= min and value.avg <= max)
 
 parseMonthlyChargesId = (value) ->
-  return null unless value? and value.raw?  
+  return null unless value.raw
```
Comment (cgrayson) [inline src/outbound.coffee]: I think some `?` protection is still be needed here, e.g., `unless value?.raw` (same on similar lines below).

File: src/outbound.coffee [hunk 4]
```diff
@@ -202,15 +191,17 @@ fullName = (first, last) ->
...
-  return null unless value?
-  if value.valueOf() then 1 else 0
+  return null unless value
+  value = types.boolean.parse(value).valueOf()
+  return 1 if value
+  return 0
 
 inRange = (value, min, max) ->
-  return false unless value?.valid
+  return false unless value.min or value.max or value.avg
```
Comment (cgrayson) [inline src/outbound.coffee]: Why did you change this? `.valid` should be a good proxy for `value` having `.min or .max or .avg`.

File: src/outbound.coffee [hunk 5]
```diff
@@ -394,31 +384,24 @@ validateRequiredId = (value, parameterName) ->
...
   lead = vars.lead
-  
-  if vars.time_in_business_id?
-    validTimeInBusinessId = validateRequiredId(vars.time_in_business_id, 'timeInBusiness') 
-  
-  if vars.average_monthly_sales_id?
-    validAverageMonthlySalesId = validateRequiredId(vars.average_monthly_sales_id, 'averageMonthlySales')
-  
-  if vars.industry_id? 
-    validIndustryId = validateRequiredId(vars.industry_id, 'industry')
```
Comment (cgrayson) [inline src/outbound.coffee]: Looks like `validateRequiredId()` can also be removed now.

File: spec/outbound-spec.coffee [hunk 6]
```diff
@@ -459,126 +417,66 @@ describe 'Outbound', ->
...
+      assert.equal error, 'company.monthly_sales must be present'  
+  
+    it 'should fail when company.industry is not present', ->
+      error = outbound.validate(variables(vars = { lead: {company: {monthly_sales: '10000-20000', industry: '', time_in_business: '20-30'}}}))
+      assert.equal error, 'company.industry must be present'    
 
-    it 'should not fail because of invalid industry_id when a valid company.industry type is present', ->
-      error = outbound.validate(idVariables({lead: {company: {industry: 'agriculture'}}}, addlVars = {lead: {lendio: {industry_id: ''}}}))
-      assert.isUndefined error
+objectVariables = () ->
```
Comment (cgrayson) [inline spec/outbound-spec.coffee]: This is pretty puzzling to me. First, I'm not getting any information from the name "objectVariables". Also, it seems that this object contains some subset of the lead data, and `variables()` contains the other half, but I can't figure out why.

File: spec/outbound-spec.coffee [hunk 7]
```diff
@@ -459,126 +417,66 @@ describe 'Outbound', ->
...
+  objectVars =
```
Comment (cgrayson) [inline spec/outbound-spec.coffee]: This local variable assignment is unnecessary.

File: src/outbound.coffee [hunk 8]
```diff
@@ -394,31 +384,24 @@ validateRequiredId = (value, parameterName) ->
...
   return 'password must not be blank' unless vars.lendio.password
   return 'first_name must not be blank' unless lead.first_name
   return 'last_name must not be blank' unless lead.last_name
-  return 'email must be a valid address' unless lead.email and lead.email.valid
-  return 'phone_1 must be a valid phone number' unless lead.phone_1.valid
-  return 'time_in_business or company.time_in_business must be present' unless lead.company.time_in_business or lead.lendio.time_in_business_id
-  return 'average_monthly_sales or company.monthly_sales must be present and valid' unless lead.company.monthly_sales or lead.lendio.average_monthly_sales_id
-  return 'industry or company.industry must be present' unless lead.company.industry or lead.lendio.industry_id
+  return 'email must not be blank' unless lead.email
+  return 'phone_1 must not be blank' unless lead.phone_1
```
Comment (cgrayson) [inline src/outbound.coffee]: These two changes are probably from `email` and `phone_1` coming in as `string` types now, but this is a regression in behavior. I.e., a phone number of `donkey` will now pass validation and be sent to Lendio. In cases like this, the integration should parse the strings into local `email` and `phone` variables, and continue to test `.valid` on those.

File: spec/inbound-spec.coffee [hunk 9]
```diff
@@ -91,7 +79,15 @@ describe 'Inbound', ->
...
-    
+
+      it 'should have loan.collateral', ->
+        assert.deepEqual integration.request(req(body())).loan.collateral, 'purchase order,other'
+
+      it 'should have loan.amount', ->
+        assert.deepEqual integration.request(req(body())).loan.amount, '5001-25000'
+
+      it 'should have loan.timeframe_days', ->
+        assert.deepEqual integration.request(req(body())).loan.timeframe_days, '1-30'
```
Comment (cgrayson) [inline spec/inbound-spec.coffee]: These 3 tests (and line 36) can use just `assert.equal`.

File: src/inbound.coffee [hunk 10]
```diff
@@ -136,6 +136,163 @@ getLastName = (value) ->
...
+  return null unless value
+  switch value
+    when '878' then '0-499'
+    when '880' then '500-599'
+    when '101' then '600-649'
+    when '100' then '650-679'  
+    when '99'  then '680-719'
+    when '98'  then '720'
+    else null
+
```
Comment (cgrayson) [inline src/inbound.coffee]: All of these new `get…()` functions are okay, but there's a kind of code duplication with all the `parse…()` functions in `outbound.coffee`. The downside is that all future changes to these codes and their values will need to be made in both places. They're essentially lookup tables, and I think they could be implemented as arrays of values, with a function that does the translation (in either direction).

File: src/outbound.coffee (hunk 1 above)
Comment (ros3bud) [inline src/outbound.coffee]: Ok, got it.

File: src/outbound.coffee (hunk 2 above)
Comment (ros3bud) [inline src/outbound.coffee]: Ay. I ran some tests and mistakenly thought it wasn't working. But I see now that the old way does work.

File: src/outbound.coffee (hunk 3 above)
Comment (ros3bud) [inline src/outbound.coffee]: Yes, I see that.

File: src/outbound.coffee (hunk 4 above)
Comment (ros3bud) [inline src/outbound.coffee]: Why? Because this integration is messing with my mind. But, yes, I see your point.

File: src/outbound.coffee (hunk 5 above)
Comment (ros3bud) [inline src/outbound.coffee]: Indeed.

File: spec/outbound-spec.coffee (hunk 6 above)
Comment (ros3bud) [inline spec/outbound-spec.coffee]: I will ping you about this. It has to do with wanting to build certain standard data types here, aka turn them into objects. But perhaps a quick conversation would be helpful.

File: src/outbound.coffee (hunk 8 above)
Comment (ros3bud) [inline src/outbound.coffee]: Ok, this makes sense.

File: spec/inbound-spec.coffee (hunk 9 above)
Comment (ros3bud) [inline spec/inbound-spec.coffee]: Ok

File: src/inbound.coffee (hunk 10 above)
Comment (ros3bud) [inline src/inbound.coffee]: Yeah, I see what you mean about that. I'd like to discuss a bit more though if it's alright.

**Comment by ros3bud** [comment]
I need to account for strings coming through for credit score.

**Comment by cgrayson** [comment]
> I need to account for strings coming through for credit score.

…? Does this mean this PR isn't ready for review yet, after all?

**Comment by ros3bud** [comment]
I just meant that I was aware it needed to be done and that I would include it in the next round after getting your feedback - however, I see that you have not yet reviewed so I can add it now.

File: src/outbound.coffee [hunk 11]
```diff
@@ -166,7 +151,47 @@ response.variables = ->
...
+    return bestKey if value.max >= highValue
+
+parseIdFromInvalidRange = (range, associativeArray) ->
+  re = new RegExp(range.raw.toLowerCase().replace(/[^a-z0-9 :]/g, ""))
+  for k, v of common.rangeValues[associativeArray]
+    return k if v.match(re)
+      
+parseIdFromString = (value, associativeArray) ->
+  return null unless value
+  value = value.toLowerCase().replace(/[^a-z]/g, " ")
```
Comment (cgrayson) [inline src/outbound.coffee]: I think the `.replace(…)` is supposed to remove non-alphabetic characters, but it doesn't.

File: src/outbound.coffee [hunk 12]
```diff
@@ -166,7 +151,47 @@ response.variables = ->
...
+  id = _.invert(common.stringValues[associativeArray])[value]
+
+parseMultipleIdsFromString= (value, associateiveArray) ->
```
Comment (cgrayson) [inline src/outbound.coffee]: Maybe there's a better name than `associativeArray` for these. Something like `lookupName`, for example, describes the purpose and not just the data structure (and is less likely to misspell :wink:).

Comment (ros3bud) [inline src/outbound.coffee]: Got it!

File: src/outbound.coffee [hunk 13]
```diff
@@ -166,7 +151,47 @@ response.variables = ->
...
+  return null unless value
+  value = value.toLowerCase().replace(/[^a-z]/g, "")
+  ids = []
+  for k, v of common.stringValues[associateiveArray]
+    re = new RegExp(v.split(' ').join('|'))
```
Comment (cgrayson) [inline src/outbound.coffee]: I don't follow the purpose of the `.split().join()` here? Looks like it would treat, for example, "assets luxury art" as `/assets|luxury|art/`, but I don't see why that's needed.

File: src/outbound.coffee [hunk 14]
```diff
@@ -166,7 +151,47 @@ response.variables = ->
...
+    highValue = 0
+    for k, v of common.rangeValues[associativeArray]
+      range = types.range.parse(v) 
+      if range.min > highValue
+        highValue = range.min
+        bestKey = k
+      return k if inRange(value, range.min, range.max)   
+    return bestKey if value.max >= highValue
+
+parseIdFromInvalidRange = (range, associativeArray) ->
```
Comment (cgrayson) [inline src/outbound.coffee]: Assuming this is used for "ranges" like [this one](https://github.com/activeprospect/leadconduit-integration-lendio/blob/standardize-inbound-lw/src/common.coffee#L49), couldn't `parseIdFromString()` just be used instead?

File: src/outbound.coffee [hunk 15]
```diff
@@ -166,7 +151,47 @@ response.variables = ->
 #
 # Helpers ----------------------------------------------------------------
 #
+# Note: time_in_business is range time in months
```
Comment (cgrayson) [inline src/outbound.coffee]: This comment looks like it might be a stray…?

File: src/outbound.coffee (hunk 11 above)
Comment (ros3bud) [inline src/outbound.coffee]: Yes, it is supposed to and I think it actually is. But upon looking into this more I see that it's taking out the strange character and adding a space. So what I need to do is make sure those extra spaces get removed, too. I'll add a better test for this to show you what I mean.

File: src/outbound.coffee (hunk 13 above)
Comment (ros3bud) [inline src/outbound.coffee]: I was using the '|' as an or. This way, a match occurs if any of the values are present. Does that make sense?

File: src/outbound.coffee (hunk 15 above)
Comment (ros3bud) [inline src/outbound.coffee]: It's actually not a stray!  I wanted to leave that information in because it was a comment on the old method: 'parseTimeInBusinessId()'. However, maybe that is no longer important because we are using the lookup tables. What do you think?

File: spec/outbound-spec.coffee [hunk 16]
```diff
@@ -459,126 +437,73 @@ describe 'Outbound', ->
...
+    lead: fields.buildLeadVars
+      credit_score: '800-900'
+      mobile_opt_in: true
+      company:
+        monthly_sales:'2000-3000'
+        state: 'Texas'
+        accounts_receivable: '22000'
+        monthly_charges: '3000-3001'
+        time_in_business: '3-5'
+        monthly_sales: '2000-3000'
```
Comment (cgrayson) [inline spec/outbound-spec.coffee]: Duplicate `monthly_sales` here.

Comment (ros3bud) [inline spec/outbound-spec.coffee]: Ok.

File: src/outbound.coffee (hunk 14 above)
Comment (ros3bud) [inline src/outbound.coffee]: That's a good catch and yes it can.

File: src/common.coffee [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+       '914': 'receivables'
+       '916': 'purchase order'
+       '918': 'inventory'
+       '920': 'other'     
+       '922': 'collateral'
+
+  primaryCustomer:
+      '787': 'business to business'
+      '788': 'business to consumer'
+      '789': 'business to government'
```
Comment (cgrayson) [inline src/common.coffee]: I like the 2-space indents on lines 150-155 a lot more than these 4-space ones. :smile:

Comment (ros3bud) [inline src/common.coffee]: Do ya now? :)

File: src/outbound.coffee [hunk 18]
```diff
@@ -166,7 +151,47 @@ response.variables = ->
...
+#
+
+
+inRange = (value, min, max) ->
+  return false unless value?.valid
```
Comment (cgrayson) [inline src/outbound.coffee]: Why check for `.valid` here, but not in `parseIdFromRange()`?

File: src/outbound.coffee [hunk 19]
```diff
@@ -166,7 +151,47 @@ response.variables = ->
 #
 # Helpers ----------------------------------------------------------------
 #
+# Note: time_in_business is range time in months
+#
+
+
+inRange = (value, min, max) ->
```
Comment (cgrayson) [inline src/outbound.coffee]: This would be simpler & clearer if it took a `range` instead of `min` and `max`.

File: src/outbound.coffee (hunk 18 above)
Comment (ros3bud) [inline src/outbound.coffee]: Because inRange needs a valid range to work properly, whereas parseIdFromRange  needs to handle an invalid range.

File: src/outbound.coffee [hunk 20]
```diff
@@ -166,7 +151,47 @@ response.variables = ->
...
+    return parseIdFromInvalidRange(range, associativeArray)
+  else
+    highValue = 0
+    for k, v of common.rangeValues[associativeArray]
+      range = types.range.parse(v) 
+      if range.min > highValue
+        highValue = range.min
+        bestKey = k
+      return k if inRange(value, range.min, range.max)   
+    return bestKey if value.max >= highValue
```
Comment (cgrayson) [inline src/outbound.coffee]: This search for the range ID is too complex to not have any comments or more test coverage.

File: src/outbound.coffee (hunk 15 above)
Comment (cgrayson) [inline src/outbound.coffee]: It just made me scratch my head, since it's not near any declaration or usage of `time_in_business`.

File: src/outbound.coffee (hunk 18 above)
Comment (cgrayson) [inline src/outbound.coffee]: Hmm, I see. Okay, that's fine defensive code, then. But… see (A) and (B).

File: src/outbound.coffee [hunk 21]
```diff
@@ -166,7 +151,47 @@ response.variables = ->
...
+#
+
+
+inRange = (value, min, max) ->
+  return false unless value?.valid
+  (value.min >= min and value.min <= max) or (value.max >= min and value.max <= max)
 
+parseIdFromRange = (value, associativeArray) ->
+  return null unless value
+  range = types.range.parse(value)
```
Comment (cgrayson) [inline src/outbound.coffee]: (A) The incoming `value` is always already a `range` type, is this needed?

File: src/outbound.coffee (hunk 15 above)
Comment (ros3bud) [inline src/outbound.coffee]: I understand what you mean - will try to think of something more clear/determine if it really needs to be here.

File: src/outbound.coffee [hunk 22]
```diff
@@ -166,7 +151,47 @@ response.variables = ->
...
+  if range.valid is false
+    return parseIdFromInvalidRange(range, associativeArray)
+  else
+    highValue = 0
+    for k, v of common.rangeValues[associativeArray]
+      range = types.range.parse(v) 
+      if range.min > highValue
+        highValue = range.min
+        bestKey = k
+      return k if inRange(value, range.min, range.max)
```
Comment (cgrayson) [inline src/outbound.coffee]: (B) This is passing `value`, even though `range` was set above as the range-parsed version of this, errr… Oh. Oh my. This is using a different `range` than above. This is really blowing my mind now.

Comment (ros3bud) [inline src/outbound.coffee]: I definitely have some variable names to update.

File: src/outbound.coffee (hunk 21 above)
Comment (ros3bud) [inline src/outbound.coffee]: I see what you mean, and now that I look at this some more, I see that it's not. I think removing that first parse() will help make the rest of the method more clear.

File: src/outbound.coffee (hunk 20 above)
Comment (cgrayson) [inline src/outbound.coffee]: …Adding: you could include this function in the `module.exports`, expressly for the purpose of having access to it from the tests and being able to test it directly. Then it can be tested directly with whatever edge cases are necessary to make sure it's right, without the unnecessary overhead of the rest of the request body, etc.

Comment (ros3bud) [inline src/outbound.coffee]: Ok, I will look into that.

File: src/inbound.coffee [hunk 23]
```diff
@@ -32,22 +33,18 @@ request = (req) ->
...
-    loan_purpose_id: lead.loan_purpose
-    time_to_funding_id: lead.time_to_funding
-    credit_score_id: lead.credit_score
-
-  company: 
+    
+  company:
+    accounts_receivable: common.rangeValues.accountsRecievable[lead.accounts_receivable_amount] or null
+    monthly_charges: common.rangeValues.monthlyCharges[lead.monthly_charges] or null
+    monthly_sales: common.rangeValues.averageMonthlySales[lead.average_monthly_sales] or null
```
Comment (cgrayson) [inline src/inbound.coffee]: This approach works, but it's a little verbose and exposes some details about `common` that it might not need to. Although it's little more than an object lookup, did you think about making these functions, similar to how they are on the outbound side? I.e., pass in the value, and the name of the lookup object, and use what it gives back? Then the `or null`s could also be built into that function, etc.

File: src/outbound.coffee (hunk 15 above)
Comment (ros3bud) [inline src/outbound.coffee]: Ok, taking it out.

File: src/inbound.coffee (hunk 23 above)
Comment (ros3bud) [inline src/inbound.coffee]: Ok, I've come around to this idea.

File: spec/outbound-spec.coffee [hunk 24]
```diff
@@ -459,126 +453,73 @@ describe 'Outbound', ->
...
-    it 'should not fail because of invalid company.time_in_business range when a valid time_in_business_id is present', ->
-      error = outbound.validate(idVariables(vars = {lead: {company: {time_in_business: 'abc'}}}))
-      assert.isUndefined error
-
-    it 'should not fail because of invalid time_in_buisness_id when a valid average_monthly_sales range is present', ->
-      error = outbound.validate(idVariables(vars = {lead: {lendio: {average_monthly_sales_id: ''}}}, addlVars = { lead: {company: {time_in_business: '30-50'}}}))
-      assert.isUndefined error 
+    it 'should fail when company.time_in_business is not present', ->
+      error = outbound.validate(variables(vars = { lead: {company: {time_in_business: ''}}}))
+      assert.equal error, 'company.time_in_business must be present'
```
Comment (cgrayson) [inline spec/outbound-spec.coffee]: Neither the `vars = ` nor any of the curly braces are needed here; the following is equivalent:
```coffeescript
      error = outbound.validate(variables(lead: company: time_in_business: '')) 
```
(Same applies to all the tests in this `describe()`).

File: spec/outbound-spec.coffee [hunk 25]
```diff
@@ -459,126 +453,73 @@ describe 'Outbound', ->
...
+    it 'should fail when company.industry is not present', ->
+      error = outbound.validate(variables(vars = { lead: {company: {monthly_sales: '10000-20000', industry: '', time_in_business: '20-30'}}}))
+      assert.equal error, 'company.industry must be present' 
 
-    it 'should not fail because of invalid industry_id when a valid company.industry type is present', ->
-      error = outbound.validate(idVariables({lead: {company: {industry: 'agriculture'}}}, addlVars = {lead: {lendio: {industry_id: ''}}}))
-      assert.isUndefined error
+  
+objectVariables = () ->
+    lead: fields.buildLeadVars
```
Comment (cgrayson) [inline spec/outbound-spec.coffee]: Our style is a 2-space indent, not 4.

File: src/inbound.coffee [hunk 26]
```diff
@@ -136,6 +133,18 @@ getLastName = (value) ->
...
+  return null unless value
+  valuesArray = []
+  for k, v of common.lookupValues[lookupName]
+    parsedKey = parseInt(k)
+    valuesArray.push v if parsedKey in value
+  if valuesArray.length then valuesArray.join(',') else null
+
+parseSingleValue = (id, lookupName) ->
+  return null unless id
+  return common.lookupValues[lookupName][id] or null
```
Comment (cgrayson) [inline src/inbound.coffee]: Pointing out only because I saw this in your very latest build-fixing commit: `return` is not needed here.

File: src/outbound.coffee [hunk 27]
```diff
@@ -14,15 +16,16 @@ _s = require('underscore.string')
 #
 # Returns a request object with the url, method, headers, and body properties.
 #
+# value for time_to_funding must be a range of days
```
Comment (cgrayson) [inline src/outbound.coffee]: Remember this?

File: src/common.coffee [hunk 28]
```diff
@@ -0,0 +1,163 @@
...
+    '110': '3-5'
+    '111': '6-11'
+    '112': '12-24'
+    '862': '25-36'
+    '864': '37-110'
+    '183': '111'
+    '109': 'not yet in business still in planning stages'
+    '944': 'looking to buy an existing business'
+
+  accountsRecievable:
```
Comment (cgrayson) [inline src/common.coffee]: Typo: `accountsReceivable`.

File: src/outbound.coffee [hunk 29]
```diff
@@ -166,8 +151,42 @@ response.variables = ->
...
+        highValue = lookUpValue.min
+        bestKey = k
+      return k if inRange(value, lookUpValue)   
+    return bestKey if value.max >= highValue
+    return null
+
+parseIdFromString = (value, lookupName) ->
+  return null unless value
+  value = value.toLowerCase().replace(/[^a-z0-9 :]/g, "")
+  id = _.invert(common.lookupValues[lookupName])[value] or null
```
Comment (cgrayson) [inline src/outbound.coffee]: No need to assign to this local variable on the last, returning line of the function.

File: src/inbound.coffee [hunk 30]
```diff
@@ -136,6 +133,18 @@ getLastName = (value) ->
   return null unless value
   value = value.split(' ').pop()
```
Comment (cgrayson) [inline src/inbound.coffee]: Noticed this could also do without the assignment in the (implicit) return statement (in `getFirstName()`, too).

File: src/common.coffee [hunk 31]
```diff
@@ -0,0 +1,163 @@
...
+    '918': 'inventory'
+    '920': 'other'     
+    '922': 'collateral'
+
+  primaryCustomer:
+    '787': 'business to business'
+    '788': 'business to consumer'
+    '789': 'business to government'
+
+  stringCreditScores:
```
Comment (cgrayson) [inline src/common.coffee]: I'm unclear on this value translation, I don't see it used on the inbound side, and `lead.credit_score` is a `range` type. Help me understand how an outbound lead would have a value like 'good' (as in the edge case tests)?

File: spec/outbound-spec.coffee (hunk 24 above)
Comment (ros3bud) [inline spec/outbound-spec.coffee]: Ok

File: spec/outbound-spec.coffee (hunk 25 above)
Comment (ros3bud) [inline spec/outbound-spec.coffee]: Ok.

File: src/inbound.coffee (hunk 26 above)
Comment (ros3bud) [inline src/inbound.coffee]: Ok, makes sense.

File: src/outbound.coffee (hunk 27 above)
Comment (ros3bud) [inline src/outbound.coffee]: Yeah, that was actually a different comment that slipped through, too. Removing.

File: src/common.coffee (hunk 28 above)
Comment (ros3bud) [inline src/common.coffee]: Ack, ok.

File: spec/outbound-spec.coffee [hunk 32]
```diff
@@ -149,47 +149,42 @@ describe 'Outbound', ->
...
+          lead: fields.buildLeadVars
+            credit_score: '1000'
+            company:
+              monthly_sales: '500000'
+            loan:
+              amount: '2000001'
+        lead = val.lead
+        creditScore = outbound.parseIdFromRange(lead.credit_score, 'rangeCreditScores')
+        averageMonthlySales = outbound.parseIdFromRange(lead.company.monthly_sales, 'averageMonthlySales')
+        financeAmounts = outbound.parseIdFromRange(lead.loan.amount, 'financeAmounts')
```
Comment (cgrayson) [inline spec/outbound-spec.coffee]: Since this is directly testing `parseIdFromRange()`, having three assertions doesn't test any more code paths than having one. (It tests three lookup objects, but I don't think that's the point here.)

File: src/outbound.coffee (hunk 29 above)
Comment (ros3bud) [inline src/outbound.coffee]: Ok.

File: src/inbound.coffee (hunk 30 above)
Comment (ros3bud) [inline src/inbound.coffee]: Ok, np.

File: src/outbound.coffee [hunk 33]
```diff
@@ -166,8 +151,42 @@ response.variables = ->
...
+  else
+    highValue = 0
+    for k, v of common.lookupValues[lookupName]
+      lookUpValue = types.range.parse(v) 
+      if lookUpValue.min > highValue
+        highValue = lookUpValue.min
+        bestKey = k
+      return k if inRange(value, lookUpValue)   
+    return bestKey if value.max >= highValue
+    return null
```
Comment (cgrayson) [inline src/outbound.coffee]: What I meant before, about this function being too complex to not have any comments or more test coverage, is that it's unclear to me what cases this code is intended to handle. Ideally the code itself is clear; comments can help explain that further as needed; tests can be consulted to show & prove what cases are handled, and how. Some of the questions I have, after studying this:

1. what is `highValue` keeping track of? 
2. why is `bestKey` better than other, lesser keys?
3. why is `bestKey` only returned when the range-being-parsed's `max` value is greater than `highValue`? 
4. what happens when the range being parsed straddles two entries in the lookup (e.g., what if `averageMonthlySales` were "25,000 - 35,000")? 
5. what if that range straddles more than two lookup entries (e.g., `averageMonthlySales` is "25,000 - 100,000")?
5. what happens when the range being parsed is lower than the lowest entry? (Most start at 0, but what if `monthlyCharges` is "500"?)

Comment (cgrayson) [inline src/outbound.coffee]: …All that said :confused:, I see that all the [code on master](https://github.com/activeprospect/leadconduit-integration-lendio/blob/master/src/outbound.coffee#L212-L222) did was a straight match on the ranges, period. Which is maybe what's happening here? 

We knew the range lookup would be more complex, and that's a problem with complex code: it's complex. :smiley: This function _looks_ like it's trying to handle some of those cases I got carried away with just above, but it doesn't. There might be some simplification here still (moving the `return k…` line to just after line 168, for example), but I think a comment to the effect that this function is only looking for simple range matches would go a long way to helping future readers.

File: spec/outbound-spec.coffee (hunk 32 above)
Comment (ros3bud) [inline spec/outbound-spec.coffee]: Well, the point is to show that the most error prone aspect of this is not getting an error. So, ok, I can see how just one is fine.

File: src/common.coffee (hunk 31 above)
Comment (ros3bud) [inline src/common.coffee]: There is overlap for the range values and strings which the credit score ID represents. For example "880: (500 - 599) Not so Good". This accounts for the possibility that someone may send either 500-599 or 'Not so good'. This also attempts to duplicate the functionality in production where either a range or string could return a particular key. The reason I did not account for this on the inbound side is because with the correct value being set based on the ID, the translation will work for outbound. I saw it as somewhat arbitrary whether I chose the string or the range for the translation, but figured that the range carries a bit more precise information.

Comment (cgrayson) [inline src/common.coffee]: Gotcha, thanks.

File: src/outbound.coffee (hunk 13 above)
Comment (cgrayson) [inline src/outbound.coffee]: Following up on this from last week again, I see now that the [code on master](https://github.com/activeprospect/leadconduit-integration-lendio/blob/master/src/outbound.coffee#L369-L380) uses regexes for _all_ the "string" lookups. 

So, a different suggestion: :smile: it would be even more clear to populate the lookup array with regexes in the first place: then there's a tiny bit less processing per lead, but more importantly the way the lookup tables are used is more obvious.

File: src/outbound.coffee (hunk 33 above)
Comment (cgrayson) [inline src/outbound.coffee]: Apologies for not being clear there, but I didn't really want you to explain the logic to me in my own comment (editing my comment is weird, btw, please don't do that again). What I want is for the code itself to be clearly understandable, with the help of comments and test coverage. For example, is there another name for `bestKey` that conveys that "it represents the key associated with the highest value"?

File: spec/outbound-spec.coffee [hunk 34]
```diff
@@ -149,55 +149,56 @@ describe 'Outbound', ->
...
+        monthlyCharges = outbound.parseIdFromRange(val.lead.company.monthly_charges, 'monthlyCharges')
+        assert.equal monthlyCharges, null
+
+      it 'should return null when value straddles multiple ranges', ->
+        val =
+          lead: fields.buildLeadVars
+            company:
+              monthly_sales: '25000-100000'
+        monthlySales = outbound.parseIdFromRange(val.lead.company.monthly_sales, 'monthlySales')
+        assert.equal monthlySales, null
```
Comment (cgrayson) [inline spec/outbound-spec.coffee]: These three `parseIdFromRange()` tests are good, I think they help a lot. A couple minor things: could inline the value of `val` (e.g., `monthlyCharges = outbound.parseIdFromRange(lead: company: monthly_charges: 'monthlyCharges')`), and these three tests don't seem to fit with the others, since they don't use any of the stuff from the `beforeEach()`.

**Comment by ros3bud** [comment]
Further concerns addressed, @cgrayson.

File: src/outbound.coffee [hunk 35]
```diff
@@ -166,8 +151,50 @@ response.variables = ->
...
+  return false unless value?.valid
+  (value.min >= range.min and value.min <= range.max) or (value.max >= range.min and value.max <= range.max)
+
+# 
+# This function takes a range and looks for a simple match 
+# based on what is available in the specified associative array within the lookup object.
+#
+# highValue is keeping track of the highest value in the specified associative array.
+# 
+# highValueKey is returned when the value.max is higher than the highest value in the specified associative array.
```
Comment (cgrayson) [inline src/outbound.coffee]: These comments look good, and help explain this logic. But often these block comments are easy to ignore; I think these notes would be even more useful down in the code. I'll demonstrate what I mean below.

File: src/outbound.coffee [hunk 36]
```diff
@@ -166,8 +151,50 @@ response.variables = ->
...
+  return null unless value
+  if value.valid is false
+    return parseIdFromString(value.raw, lookupName) or null
+  else
+    highValue = 0
+    for k, v of common.lookupValues[lookupName]
+      lookUpValue = types.range.parse(v) 
+      return k if inRange(value, lookUpValue) 
+      if lookUpValue.min > highValue
+        highValue = lookUpValue.min
```
Comment (cgrayson) [inline src/outbound.coffee]: Right here where it's assigned is a good place to explain what it's for: 

```coffeescript
      highValue = lookUpValue.min # keep track of the highest value in the lookup array
```

File: src/outbound.coffee [hunk 37]
```diff
@@ -166,8 +151,50 @@ response.variables = ->
...
+        highValueKey = k  
+    return highValueKey if value.max >= highValue
```
Comment (cgrayson) [inline src/outbound.coffee]: And right above this return, for example:

```
    # we didn't find a match, but if value is higher than highest found, return that key
```

**Comment by cgrayson** [comment]
Otherwise… 

lgtm borat

File: src/outbound.coffee [hunk 38]
```diff
@@ -166,8 +151,50 @@ response.variables = ->
...
+    return null
+
+parseIdFromString = (value, lookupName) ->
+  return null unless value
+  value = value.toLowerCase().replace(/[^a-z0-9 :]/g, "")
+  _.invert(common.lookupValues[lookupName])[value] or null
```
Comment (alexkwolfe) [inline src/outbound.coffee]: Instead of calling `_.invert` for every lead, why not call it once in the common file? Looping over arrays is bad for performance in an event reactor because during that loop, nothing else can be done. This is a simple way to avoid unnecessary looping.

For example:

```coffeescript
codeCollateral = 
  '946': 'assets luxury art'
  '904': 'equipment trucks medical'
  '906': 'cash investments'
  '924': 'retirement account 401k'
  
collateralCode = _.invert(codeCollateral)
```
